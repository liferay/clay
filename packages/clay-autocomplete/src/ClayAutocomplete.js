import 'clay-data-provider';
import {Config} from 'metal-state';
import {isFunction} from 'metal';
import ClayComponent from 'clay-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import templates from './ClayAutocomplete.soy.js';

/**
 * Metal ClayAutocomplete component.
 * @extends ClayComponent
 */
class ClayAutocomplete extends ClayComponent {
	/**
	 * @inheritDoc
	 */
	attached() {
		this._dropdownItemFocused = null;

		this.addListener('dataChange', this._defaultDataChange, true);
		this.addListener('inputChange', this._defaultInputChange, true);
	}

	/**
	 * @inheritDoc
	 */
	disposed() {
		this._dropdownItemFocused = null;
	}

	/**
	 * @inheritDoc
	 */
	syncFilteredItems() {
		this._dropdownItemFocused = null;
	}

	/**
	 * Sets filtered items with received data
	 * @private
	 */
	_defaultDataChange() {
		if (this._query) {
			this.filteredItems = this.refs.dataProvider.filter(
				this._query,
				this.extractData
			);
		} else {
			this.filteredItems = [];
		}
	}

	/**
	 * Filters the items according to received input
	 * @param {!Event} event
	 * @private
	 */
	_defaultInputChange(event) {
		this._query = event.data.value;

		if (this._query) {
			if (isFunction(this.dataSource)) {
				this.refs.dataProvider.updateData(this._query);
			} else {
				this.filteredItems = this.refs.dataProvider.filter(
					this._query,
					this.extractData
				);
			}
		} else {
			this.filteredItems = [];
		}
	}

	/**
	 * Continues the propagation of the data change event
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleDataChange(event) {
		return !this.emit({
			data: event.data,
			name: 'dataChange',
			originalEvent: event,
		});
	}

	/**
	 * Continues the propagation of the item clicked event
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleDropdownItemClick(event) {
		return this._handleItemSelected(event);
	}

	/**
	 * Handle the selected item in the dropdown and trigger the itemSelected event.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleItemSelected(event) {
		const index = event.target.getAttribute('data-dropdown-item-index');
		const item = this.filteredItems[Number(index)];

		return !this.emit({
			data: item.data,
			name: 'itemSelected',
			originalEvent: event,
		});
	}

	/**
	 * Handles the input blur event and continues to propagate.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleOnBlur(event) {
		return !this.emit({
			name: 'inputOnBlur',
			originalEvent: event,
		});
	}

	/**
	 * Handles the input focus event and continues to propagate.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleOnFocus(event) {
		return !this.emit({
			name: 'inputOnFocus',
			originalEvent: event,
		});
	}

	/**
	 * Handles the input event and continues to propagate.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleOnInput(event) {
		return !this.emit({
			data: {
				value: event.target.value,
				char: event.data,
			},
			name: 'inputChange',
			originalEvent: event,
		});
	}

	/**
	 * Handles the onKeydown events and continues to propagate.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleOnKeydown(event) {
		switch (event.key) {
		case 'Enter':
			event.stopPropagation();
			if (
				this.filteredItems.length &&
					event.delegateTarget.tagName !== 'INPUT'
			) {
				this._handleItemSelected(event);
			}
			break;
		case 'ArrowUp':
			this._setFocusItem(true);
			break;
		case 'ArrowDown':
			event.stopPropagation();
			this._setFocusItem(false);
			break;
		}

		return !this.emit({
			data: {
				value: this.refs.input.value,
				key: event.key,
				eventFromInput: event.delegateTarget.tagName === 'INPUT',
			},
			name: 'inputOnKeydown',
			originalEvent: event,
		});
	}

	/**
	 * Handle the interactions in the dropdown and add focus on the items.
	 * @param {!Boolean} direction
	 * @protected
	 */
	_setFocusItem(direction) {
		if (this.filteredItems.length) {
			const {dropdown} = this.refs.dataProvider.refs;
			const elements = dropdown.querySelectorAll(
				'a[data-dropdown-item-index]'
			);

			if (direction && this._dropdownItemFocused === 0) {
				this.refs.input.focus();
				this._dropdownItemFocused = null;
			} else if (
				!(direction === true && this._dropdownItemFocused === null)
			) {
				this._dropdownItemFocused =
					this._dropdownItemFocused === null ||
					elements.length - 1 === this._dropdownItemFocused
						? direction
							? this._dropdownItemFocused - 1
							: 0
						: direction
							? this._dropdownItemFocused - 1
							: this._dropdownItemFocused + 1;

				elements[this._dropdownItemFocused].focus();
			}
		}
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayAutocomplete.STATE = {
	/**
	 * Variation name to render different deltemplates.
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(string|undefined)}
	 */
	contentRenderer: Config.string(),

	/**
	 * Data to add to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?object}
	 */
	data: Config.object(),

	/**
	 * The array of data items that the data source contains,
	 * the URL for the data provider to request, or a function
	 * that receives the query and returns a promise with the
	 * elements.
	 *
	 * @instance
	 * @default undefined
	 * @memberof ClayAutocomplete
	 * @type {!(string|object|array|function)}
	 */
	dataSource: Config.oneOfType([
		Config.array(),
		Config.func(),
		Config.object(),
		Config.string(),
	]).required(),

	/**
	 * Set the request debounce time
	 * @instance
	 * @default 200
	 * @memberof ClayDataProvider
	 * @type {?(number)}
	 */
	debounceTime: Config.number().value(200),

	/**
	 * Object that wires events with default listeners
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @review
	 * @type {?(object|undefined)}
	 */
	defaultEventHandler: Config.object(),

	/**
	 * CSS classes to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(string|undefined)}
	 */
	elementClasses: Config.string(),

	/**
	 * Flag to enable the filtred items with autocomplete default.
	 * @default true
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?bool}
	 */
	enableAutocomplete: Config.validator(value => {
		if (value) {
			console.warn(
				'ðŸš¨ `enableAutocomplete` has been deprecated and will be removed in the next major version.'
			);
		}
	}),

	/**
	 * Extracts from the data the item to be compared in autocomplete.
	 * @instance
	 * @default (elem) => elem
	 * @memberof ClayAutocomplete
	 * @type {?(function|string)}
	 */
	extractData: Config.oneOfType([Config.func(), Config.string()]).value(
		elem => elem
	),

	/**
	 * List of filtered items for suggestion or autocomplete.
	 * @default []
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?Array<Object>}
	 */
	filteredItems: Config.array(Config.object()).value([]),

	/**
	 * Id to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(string|undefined)}
	 */
	id: Config.string(),

	/**
	 * Set some initial data while the first request is being made
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(object|array)}
	 */
	initialData: Config.oneOfType([Config.object(), Config.array()]),

	/**
	 * CSS classes to be applied to the input element.
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(string|undefined)}
	 */
	inputElementClasses: Config.string(),

	/**
	 * Name of the selected items input.
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(string|undefined)}
	 */
	inputName: Config.string(),

	/**
	 * Value of the input.
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(string|undefined)}
	 */
	inputValue: Config.string(),

	/**
	 * Flag to define how often to refetch data (ms)
	 * @instance
	 * @default 0
	 * @memberof ClayAutocomplete
	 * @type {?(number|undefined)}
	 */
	pollingInterval: Config.number().value(0),

	/**
	 * Specifies explicitly if request needs to be made with debounce
	 * (userInput) or with polling (polling)
	 * @instance
	 * @default undefined
	 * @memberof ClayAutocomplete
	 * @type {?(object|array)}
	 */
	requestInputMode: Config.oneOf(['polling', 'userInput']).value('userInput'),

	/**
	 * Set ups the request options
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(object|undefined)}
	 */
	requestOptions: Config.shapeOf({
		method: Config.string(),
		mode: Config.string(),
		cache: Config.string(),
		credentials: Config.string(),
		headers: Config.object(),
		redirect: Config.string(),
		referrer: Config.string(),
		body: Config.object(),
	}),

	/**
	 * Flag to define how often to refetch data (ms)
	 * @instance
	 * @default 0
	 * @memberof ClayAutocomplete
	 * @type {?(number|undefined)}
	 */
	requestPolling: Config.validator(value => {
		if (value) {
			console.warn(
				'ðŸš¨ `requestPolling` has been renamed to `pollingInterval` and will be deprecated and removed in the next release.'
			);
		}
	}),

	/**
	 * Define how many attempts will be made when the request fails
	 * @default 5
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(number|undefined)}
	 */
	requestRetries: Config.number().value(5),

	/**
	 * Set timeout of the request
	 * @default 30000
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?(number|undefined)}
	 */
	requestTimeout: Config.number().value(30000),

	/**
	 * The content renderer.
	 * Warning: This API is unstable and can be changed at any time,
	 * use at your own risk.
	 * @default undefined
	 * @instance
	 * @memberof ClayAutocomplete
	 * @type {?html}
	 */
	unstable_content: Config.any(),
};

defineWebComponent('clay-autocomplete', ClayAutocomplete);

Soy.register(ClayAutocomplete, templates);

export {ClayAutocomplete};
export default ClayAutocomplete;
