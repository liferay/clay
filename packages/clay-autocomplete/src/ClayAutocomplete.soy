{namespace ClayAutocomplete}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param dataSource: string|[]|list<?>|any}
	{@param? _handleDataChange: any}
	{@param? _handleDropdownItemClick: any}
	{@param? _handleOnBlur: any}
	{@param? _handleOnFocus: any}
	{@param? _handleOnInput: any}
	{@param? _handleOnKeydown: any}
	{@param? contentRenderer: string}
	{@param? elementClasses: string}
	{@param? filteredItems: list<?>}
	{@param? id: string}
	{@param? initialData: []|list<?>}
	{@param? inputElementClasses: string}
	{@param? inputName: string}
	{@param? inputValue: string}
	{@param? requestOptions: [
		method: string,
		mode: string,
		cache: string,
		credentials: string,
		headers: [],
		redirect: string,
		referrer: string,
		body: []
	]}
	{@param? pollingInterval: number}
	{@param? requestInputMode: string}
	{@param? requestRetries: number}
	{@param? requestTimeout: number}
	{@param? unstable_content: html}

	{let $attributes kind="attributes"}
		class="dropdown-full
			{if $elementClasses}
				{sp}{$elementClasses}
			{else}
				{sp}form-group
			{/if}
		"

		data-onkeydown="{$_handleOnKeydown}"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	<div {$attributes}>
		{$unstable_content}

		{call .input}
			{param _handleOnBlur: $_handleOnBlur /}
			{param _handleOnFocus: $_handleOnFocus /}
			{param _handleOnInput: $_handleOnInput /}
			{param _handleOnKeydown: $_handleOnKeydown /}
			{param inputElementClasses: $inputElementClasses /}
			{param inputName: $inputName /}
			{param inputValue: $inputValue /}
		{/call}

		{call .dropdown}
			{param _handleDataChange: $_handleDataChange /}
			{param _handleDropdownItemClick: $_handleDropdownItemClick /}
			{param contentRenderer: $contentRenderer /}
			{param dataSource: $dataSource /}
			{param filteredItems: $filteredItems /}
			{param initialData: $initialData /}
			{param requestOptions: $requestOptions /}
			{param pollingInterval: $pollingInterval /}
			{param requestRetries: $requestRetries /}
			{param requestInputMode: $requestInputMode /}
			{param requestTimeout: $requestTimeout /}
		{/call}
	</div>
{/template}

/**
 * This render the input tag.
 */
{template .input}
	{@param? _handleOnBlur: any}
	{@param? _handleOnFocus: any}
	{@param? _handleOnInput: any}
	{@param? _handleOnKeydown: any}
	{@param? inputElementClasses: string}
	{@param? inputName: string}
	{@param? inputValue: string}

	{let $attributes kind="attributes"}
		class="
			{if $inputElementClasses}
				{$inputElementClasses}
			{else}
				form-control
			{/if}
		"

		{if $inputName}
			id="{$inputName}"
			name="{$inputName}"
		{/if}

		onBlur="{$_handleOnBlur}"
		onFocus="{$_handleOnFocus}"
		onInput="{$_handleOnInput}"
		onKeydown="{$_handleOnKeydown}"
		ref="input"
		type="text"

		{if $inputValue}
			value="{$inputValue}"
		{else}
			value=""
		{/if}
	{/let}

	<input {$attributes} />
{/template}

/**
 * This render the dropdown content.
 */
{template .dropdown}
	{@param dataSource: string|[]|list<?>|any}
	{@param? _handleDataChange: any}
	{@param? _handleDropdownItemClick: any}
	{@param? contentRenderer: string}
	{@param? filteredItems: list<?>}
	{@param? initialData: []|list<?>}
	{@param? requestOptions: [
		method: string,
		mode: string,
		cache: string,
		credentials: string,
		headers: [],
		redirect: string,
		referrer: string,
		body: []
	]}
	{@param? pollingInterval: number}
	{@param? requestInputMode: string}
	{@param? requestRetries: number}
	{@param? requestTimeout: number}

	{let $isVisible: $filteredItems and length($filteredItems) > 0 ? true : false /}

	{let $classes kind="text"}
		dropdown-menu
		{if $isVisible}
			{sp}show
		{/if}
	{/let}

	{call ClayDataProvider.render}
		{param content kind="html"}
			<ul class="{$classes}" ref="dropdown">
				{if $filteredItems}
					{foreach $item in $filteredItems}
						{delcall ClayAutocomplete.Item variant="$contentRenderer" allowemptydefault="true"}
							{param _handleDropdownItemClick: $_handleDropdownItemClick /}
							{param data: $item /}
							{param index: index($item) /}
						{/delcall}
					{/foreach}
				{/if}
			</ul>
		{/param}

		{param dataSource: $dataSource /}
		{param events: ['dataChange': $_handleDataChange] /}
		{param initialData: $initialData /}
		{param ref: 'dataProvider' /}
		{param requestOptions: $requestOptions /}
		{param pollingInterval: $pollingInterval /}
		{param requestRetries: $requestRetries /}
		{param requestInputMode: $requestInputMode /}
		{param requestTimeout: $requestTimeout /}
	{/call}
{/template}

/**
 * This renders the item content of a dropdown
 * Extension point: Create a new variation above ClayAutocomplete.Item
 * so you can create your own markup.
 */
{deltemplate ClayAutocomplete.Item}
	{@param data: ?}
	{@param index: int}
	{@param? _handleDropdownItemClick: any}

	<li>
		<a data-onclick="{$_handleDropdownItemClick}" data-dropdown-item-index="{$index}" id="item" class="dropdown-item" href="javascript:;">
			{if $data.matches and length($data.matches) > 0}
				{foreach $char in $data.matches}
					{if $char.match}
						<strong>{$char.value}</strong>
					{else}
						{$char.value}
					{/if}
				{/foreach}
			{/if}
		</a>
	</li>
{/deltemplate}