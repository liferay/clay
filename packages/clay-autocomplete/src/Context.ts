/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import React from 'react';

interface IContext {
	/**
	 * Status of a data fetch is happening, this will help
	 * <ClayAutocomplete.Input /> add necessary markups.
	 */
	loading: boolean;

	/**
	 * Changes the loading status and causes a new rendering.
	 */
	onLoadingChange: (loading: boolean) => void;

	/**
	 * Ref of the container to align the dropdown.
	 */
	containerElementRef: React.RefObject<HTMLElement>;
}

const context = React.createContext({} as IContext);

context.displayName = 'ClayAutocompleteContext';

export default context;
