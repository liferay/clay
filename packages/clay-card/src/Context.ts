/**
 * SPDX-FileCopyrightText: Â© 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */

import React from 'react';

export interface IContext {
	/**
	 * Flag to indicate whether to use card as a horizontal/folder card.
	 * See the following documenation:
	 * https://liferay.design/lexicon/core-components/cards/cards-variation/
	 */
	horizontal?: boolean;

	/**
	 * Flag to indicate whether to use card as a navigation card.
	 * See the following documenation:
	 * https://liferay.design/lexicon/core-components/cards/cards-navigation/
	 */
	interactive?: boolean;
}

const context = React.createContext({} as IContext);

context.displayName = 'ClayCardContext';

export default context;
