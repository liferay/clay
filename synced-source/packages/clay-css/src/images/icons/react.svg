<!--
* SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
* SPDX-FileCopyrightText: © 2020 Contributors to the project Clay <https://github.com/liferay/clay/graphs/contributors>
*
* SPDX-License-Identifier: BSD-3-Clause
-->
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
	<path class="lexicon-icon-outline" d="M512,259.8c0-33.9-42.5-66-107.6-85.9c15-66.3,8.3-119.1-21.1-136c-6.8-4-14.6-5.8-23.3-5.8V32c-28.7,0-66.3,20.4-104.3,55.9c-38-35.2-75.6-55.5-104.3-55.5v0.1c-8.8,0-16.7,1.9-23.6,5.8C98.6,55.3,92,107.9,107.2,174C42.3,194,0,226.1,0,259.8c0,33.9,42.5,66,107.6,85.9c-15,66.3-8.3,119.1,21.1,136c6.8,4,14.7,5.8,23.5,5.8c28.7,0,66.3-20.4,104.3-55.9c38,35.2,75.6,55.5,104.3,55.5c8.8,0,16.7-1.9,23.6-5.8c29.3-16.9,35.9-69.6,20.8-135.7C469.7,325.7,512,293.6,512,259.8z M359.9,55.3c4.8,0.1,8.7,1,11.9,2.8c14.2,8.1,20.3,39.1,15.5,78.9c-1.1,9.8-3,20.1-5.3,30.7c-20.5-5-42.8-8.9-66.3-11.4c-14.1-19.3-28.7-36.8-43.4-52.1C306.3,72.6,338.2,55.4,359.9,55.3z M163.5,313.2c4.9,8.3,9.9,16.5,15,24.4c-14.8-2.2-29.1-4.9-42.8-8.2c3.9-13.5,8.7-27.3,14.1-41.2C154.1,296.5,158.6,304.9,163.5,313.2z M135.5,190.3c13.7-3.2,28.1-6,43-8.2c-5.1,8-10.2,16.3-15,24.7c-4.8,8.3-9.3,16.7-13.6,25C144.2,217.8,139.4,203.8,135.5,190.3z M162.1,260.2c6.5-14,13.8-27.9,21.6-41.6c8.1-14.1,16.5-27.4,25.2-39.8c15.6-1.4,31.3-2.1,47.2-2.1c15.8,0,31.5,0.7,47,2c8.7,12.4,17.1,25.7,25.3,39.6c7.9,13.7,15.1,27.5,21.7,41.5c-6.6,14-13.8,27.9-21.6,41.6c-8.1,14.1-16.5,27.4-25.2,39.8c-15.5,1.4-31.3,2.1-47.2,2.1c-15.8,0-31.5-0.7-47-2c-8.7-12.4-17.1-25.7-25.3-39.6C175.9,288,168.6,274.1,162.1,260.2z M348.5,312.9c4.8-8.3,9.3-16.8,13.6-25.1c5.6,14,10.4,27.9,14.4,41.5c-13.7,3.3-28.1,6.2-43,8.3C338.5,329.6,343.7,321.4,348.5,312.9z M348.4,206.5c-4.8-8.3-9.9-16.5-15-24.4c14.8,2.2,29.1,4.9,42.8,8.2c-3.9,13.5-8.7,27.3-14.1,41.2C357.8,223.1,353.3,214.8,348.4,206.5z M284.7,154c-9.4-0.4-19-0.7-28.7-0.7c-9.8,0-19.5,0.2-29,0.7c9.4-12.2,19.1-23.4,28.7-33.4C265.4,130.6,275.1,141.8,284.7,154z M139.5,58.5c3.1-1.9,7.2-2.7,12-2.7v-0.1c21.6,0,53.6,17.2,87.7,48.6c-14.6,15.3-29.2,32.7-43.1,52c-23.6,2.5-45.9,6.4-66.4,11.5c-2.4-10.4-4.2-20.5-5.4-30.2C119.4,97.7,125.4,66.7,139.5,58.5z M113.2,323.2c-10.3-3.2-20.1-6.8-29.2-10.6c-36.9-15.7-60.8-36.4-60.8-52.8s23.9-37.1,60.8-52.8c9-3.9,18.8-7.3,28.9-10.5c5.9,20.4,13.8,41.7,23.5,63.5C126.8,281.7,119.1,302.9,113.2,323.2z M152,464.4c-4.7-0.1-8.7-1-11.8-2.8c-14.2-8.1-20.4-39.1-15.6-78.9c1.1-9.8,3-20.1,5.3-30.7c20.5,5,42.8,8.9,66.3,11.4c14.1,19.3,28.7,36.8,43.4,52.1C205.6,447.1,173.7,464.4,152,464.4z M256.2,399c-9.7-10-19.4-21.2-29-33.4c9.4,0.4,19,0.7,28.7,0.7c9.8,0,19.5-0.2,29-0.7C275.5,377.9,265.8,389,256.2,399z M372.5,461.2c-3.1,1.9-7.2,2.7-12,2.7c-21.6,0-53.6-17.2-87.7-48.6c14.6-15.3,29.2-32.7,43.1-52c23.6-2.5,45.9-6.4,66.4-11.5c2.4,10.5,4.3,20.6,5.4,30.3C392.6,422,386.6,453,372.5,461.2z M427.9,312.6c-9,3.9-18.8,7.3-28.9,10.5c-5.9-20.4-13.8-41.7-23.5-63.5c9.6-21.7,17.3-42.9,23.2-63.2c10.3,3.2,20.1,6.8,29.3,10.6c36.9,15.7,60.8,36.4,60.8,52.8C488.7,276.2,464.8,297,427.9,312.6z"/>
	<path class="lexicon-icon-outline" d="M255.9,307.5c26.3,0,47.7-21.3,47.7-47.7s-21.4-47.7-47.7-47.7c-26.3,0-47.7,21.3-47.7,47.7S229.6,307.5,255.9,307.5z"/>
</svg>