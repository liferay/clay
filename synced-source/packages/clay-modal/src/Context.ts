/**
 * SPDX-FileCopyrightText: Â© 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */

import React from 'react';

import {Status} from './types';

export interface IContext {
	ariaLabelledby?: string;

	/**
	 * Callback called to close the modal.
	 */
	onClose: () => void;

	/**
	 * The path to the SVG spritemap file containing the icons.
	 */
	spritemap?: string;

	/**
	 * Status messages.
	 */
	status?: Status;
}

const context = React.createContext({} as IContext);

context.displayName = 'ClayModalContext';

export default context;
