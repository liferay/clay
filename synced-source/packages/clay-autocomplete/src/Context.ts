/**
 * SPDX-FileCopyrightText: Â© 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */

import {createContext, useContext} from 'react';

interface IContext {
	/**
	 * Status of a data fetch is happening, this will help
	 * <ClayAutocomplete.Input /> add necessary markups.
	 */
	loading: boolean;

	/**
	 * Changes the loading status and causes a new rendering.
	 */
	onLoadingChange: (loading: boolean) => void;

	/**
	 * Ref of the container to align the dropdown.
	 */
	containerElementRef: React.RefObject<HTMLElement>;
}

export const LegacyContext = createContext({} as IContext);

LegacyContext.displayName = 'ClayAutocompleteLegacyContext';

type AutocompleteContext = {
	activeDescendant?: React.Key;
	onActiveDescendant: (value: React.Key) => void;
	onClick: (value: string) => void;
};

export const AutocompleteContext = createContext({} as AutocompleteContext);

export function useAutocompleteState() {
	return useContext(AutocompleteContext);
}
