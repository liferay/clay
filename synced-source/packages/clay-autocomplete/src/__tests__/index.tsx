/**
 * SPDX-FileCopyrightText: Â© 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */

/* eslint-disable react/display-name */
import ClayAutocomplete from '..';
import {cleanup, render} from '@testing-library/react';
import React from 'react';

import {LegacyContext} from '../Context';

describe('ClayAutocomplete', () => {
	afterEach(cleanup);

	it('renders', () => {
		const {container} = render(<ClayAutocomplete />);

		expect(container).toMatchSnapshot();
	});

	it('renders Autocomplete with other markup component', () => {
		const MyMarkup = React.forwardRef<
			HTMLDivElement,
			React.HTMLAttributes<HTMLDivElement>
		>(({children, ...otherProps}, ref) => (
			<div
				{...otherProps}
				className="form-control form-control-tag-group"
				ref={ref}
			>
				{children}
			</div>
		));

		const {container} = render(<ClayAutocomplete component={MyMarkup} />);

		expect(container).toMatchSnapshot();
	});

	it('renders Item with matches values', () => {
		const {container} = render(
			<ClayAutocomplete.Item match="Baz" value="Bar" />
		);

		expect(container).toMatchSnapshot();
	});

	it('renders LoadingIndicator', () => {
		const LoadingIndicatorWithContext = () => (
			<LegacyContext.Provider
				value={{
					containerElementRef: {current: null},
					loading: false,
					onLoadingChange: jest.fn(),
				}}
			>
				<ClayAutocomplete.LoadingIndicator />
			</LegacyContext.Provider>
		);
		const {container} = render(<LoadingIndicatorWithContext />);

		expect(container).toMatchSnapshot();
	});

	it('renders LoadingIndicator with other markup component', () => {
		const MyMarkup = ({
			children,
		}: React.HTMLAttributes<HTMLSpanElement>) => (
			<span className="autofit-col">
				<span className="inline-item">{children}</span>
			</span>
		);
		const LoadingIndicatorWithContext = () => (
			<LegacyContext.Provider
				value={{
					containerElementRef: {current: null},
					loading: false,
					onLoadingChange: jest.fn(),
				}}
			>
				<ClayAutocomplete.LoadingIndicator component={MyMarkup} />
			</LegacyContext.Provider>
		);

		const {container} = render(<LoadingIndicatorWithContext />);

		expect(container).toMatchSnapshot();
	});

	it('renders LoadingIndicator and calls the onLoadingChange', () => {
		const spyFn = jest.fn();
		const LoadingIndicatorWithContext = () => (
			<LegacyContext.Provider
				value={{
					containerElementRef: {current: null},
					loading: false,
					onLoadingChange: spyFn,
				}}
			>
				<ClayAutocomplete.LoadingIndicator />
			</LegacyContext.Provider>
		);

		render(<LoadingIndicatorWithContext />);

		expect(spyFn).toBeCalled();
		expect(spyFn).toBeCalledWith(true);
	});

	it('renders Input with classNames when loading for true', () => {
		const InputWithContext = () => (
			<LegacyContext.Provider
				value={{
					containerElementRef: {current: null},
					loading: true,
					onLoadingChange: jest.fn(),
				}}
			>
				<ClayAutocomplete.Input />
			</LegacyContext.Provider>
		);

		const {container} = render(<InputWithContext />);

		expect(container).toMatchSnapshot();
	});

	it('renders DropDown with alignment with container', () => {
		const DropDownAndInputWithContext = () => {
			const containerElementRef = React.useRef<HTMLDivElement>(null);

			return (
				<LegacyContext.Provider
					value={{
						containerElementRef,
						loading: false,
						onLoadingChange: jest.fn(),
					}}
				>
					<ClayAutocomplete.Input />
					<ClayAutocomplete.DropDown active />
				</LegacyContext.Provider>
			);
		};

		render(<DropDownAndInputWithContext />);

		expect(document.body).toMatchSnapshot();
	});
});
