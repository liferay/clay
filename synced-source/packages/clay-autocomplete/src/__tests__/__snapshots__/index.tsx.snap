// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayAutocomplete renders 1`] = `
<div>
  <div
    class="input-group"
  >
    <div
      class="input-group-item"
    />
  </div>
</div>
`;

exports[`ClayAutocomplete renders Autocomplete with other markup component 1`] = `
<div>
  <div
    class="form-control form-control-tag-group"
  />
</div>
`;

exports[`ClayAutocomplete renders DropDown with alignment with container 1`] = `
<body>
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <input
      class="form-control"
      type="text"
    />
  </div>
  <div>
    <div>
      <div
        class="dropdown-menu autocomplete-dropdown-menu show"
        role="presentation"
        style="max-width: none;"
      />
    </div>
  </div>
</body>
`;

exports[`ClayAutocomplete renders Input with classNames when loading for true 1`] = `
<div>
  <input
    class="form-control input-group-inset input-group-inset-after"
    type="text"
  />
</div>
`;

exports[`ClayAutocomplete renders Item with matches values 1`] = `
<div>
  <li
    role="presentation"
  >
    <span
      class="dropdown-item"
      role="menuitem"
    >
      Bar
    </span>
  </li>
</div>
`;

exports[`ClayAutocomplete renders LoadingIndicator 1`] = `
<div>
  <div
    class="input-group-inset-item input-group-inset-item-after"
  >
    <span
      class="inline-item inline-item-middle"
    >
      <span
        aria-hidden="true"
        class="loading-animation loading-animation-sm"
      />
    </span>
  </div>
</div>
`;

exports[`ClayAutocomplete renders LoadingIndicator with other markup component 1`] = `
<div>
  <span
    class="autofit-col"
  >
    <span
      class="inline-item"
    >
      <span
        aria-hidden="true"
        class="loading-animation loading-animation-sm"
      />
    </span>
  </span>
</div>
`;
